### PR TITLE
FIX (GraphEngine): @W-11992240@: Handle more SObjectField and DescribeSObjectResult formats

### PR DESCRIPTION
**Pull request by rmohan20**
This PR handles the following formats:
SObjectField: Account.Name, MyObject__c.MyField__c
DescribeSObjectResult: SObjectType.Account

### CODE REVIEW

File: sfge/src/main/java/com/salesforce/graph/symbols/apex/schema/DescribeSObjectResultFactory.java [hunk 1]
```diff
@@ -77,6 +85,20 @@ public final class DescribeSObjectResultFactory {
                 return Optional.empty();
             };
 
+    private static Optional<ApexValue<?>> getSObjectTypeValue(String sObjectType, String nameProperty) {
```
Comment (rmohan20) [inline sfge/src/main/java/com/salesforce/graph/symbols/apex/schema/DescribeSObjectResultFactory.java:88]: Extracted existing logic into its own method for reusability.

File: sfge/src/main/java/com/salesforce/graph/symbols/apex/schema/DescribeSObjectResultFactory.java [hunk 2]
```diff
@@ -106,33 +128,46 @@ public final class DescribeSObjectResultFactory {
...
+        final DescribeSObjectResult describeSObjectResult =
+                ApexValueBuilder.getWithoutSymbolProvider()
+                        .valueVertex(vertex)
+                        .returnedFrom(sObjectType, vertex)
+                        .buildDescribeSObjectResult(sObjectType);
+        return Optional.of(
+                ApexValueBuilder.getWithoutSymbolProvider()
+                        .returnedFrom(describeSObjectResult, vertex)
+                        .buildBoolean());
+    }
```
Comment (rmohan20) [inline sfge/src/main/java/com/salesforce/graph/symbols/apex/schema/DescribeSObjectResultFactory.java:170]: Refactor for reusability.

File: sfge/src/main/java/com/salesforce/rules/fls/apex/operations/SchemaBasedValidationAnalyzer.java [hunk 3]
```diff
@@ -93,6 +93,7 @@ public Set<FlsValidationRepresentation.Info> checkForValidation(
 
         // Convert the value to a boolean that represents if the object was initialized
         // TODO: This might need to be more specific
+        // FIXME: Handle forloop correctly
```
Comment (rmohan20) [inline sfge/src/main/java/com/salesforce/rules/fls/apex/operations/SchemaBasedValidationAnalyzer.java:96]: Change for another PR leaked here. Will remove this.

File: sfge/src/main/java/com/salesforce/graph/symbols/apex/schema/DescribeSObjectResultFactory.java [hunk 4]
```diff
@@ -15,7 +17,22 @@ public final class DescribeSObjectResultFactory {
             vertex -> {
                 List<String> chainedNames = vertex.getChainedNames();
 
-                if (chainedNames.size() == 2) {
+                if (chainedNames.size() == 1) {
+                    // SObjectType.Account
```
Comment (jfeingold35) [inline sfge/src/main/java/com/salesforce/graph/symbols/apex/schema/DescribeSObjectResultFactory.java:21]: To be sure I understand, this is when you reference something like `SObjectType.Account` without using the `Schema` namespace first? If so, could a comment be added to explain how this differs from the case at line 35?

File: sfge/src/main/java/com/salesforce/graph/symbols/apex/schema/DescribeSObjectResultFactory.java [hunk 5]
```diff
@@ -106,33 +128,46 @@ public final class DescribeSObjectResultFactory {
                                         .returnedFrom(sObjectType, vertex)
                                         .buildDescribeSObjectResult(sObjectType));
                     }
-                } else if (SystemNames.DML_OBJECT_ACCESS_METHODS.contains(methodName)
-                        && chainedNames.size() == 3) {
+                } else if (SystemNames.DML_OBJECT_ACCESS_METHODS.contains(methodName)) {
                     // Schema.SObjectType.Account.isDeletable()
-                    if (chainedNames
+                    if (chainedNames.size() == 3
```
Comment (jfeingold35) [inline sfge/src/main/java/com/salesforce/graph/symbols/apex/schema/DescribeSObjectResultFactory.java:133]: This somehow feels redundant with the else-if at line 144. Could the if-condition be widened to include both cases and the code made flexible enough to cover both?

File: sfge/src/main/java/com/salesforce/graph/symbols/apex/schema/DescribeSObjectResultFactory.java (hunk 4 above)
Comment (rmohan20) [reply sfge/src/main/java/com/salesforce/graph/symbols/apex/schema/DescribeSObjectResultFactory.java:21]: Right, that's the difference. Will add more information to the code.

File: sfge/src/main/java/com/salesforce/graph/symbols/apex/schema/DescribeSObjectResultFactory.java (hunk 5 above)
Comment (rmohan20) [reply sfge/src/main/java/com/salesforce/graph/symbols/apex/schema/DescribeSObjectResultFactory.java:133]: Yeah, handling the `Schema.` prefix on some types and lack of it has been creating some redundancy. Will see how I can make this nicer.

File: sfge/src/main/java/com/salesforce/graph/symbols/apex/schema/SObjectTypeChainUtil.java [hunk 6]
```diff
@@ -0,0 +1,98 @@
...
+import java.util.List;
+import java.util.Optional;
+
+/**
+ * Utility to help checking chained name prefixes in Schema-based invocations. For now, this only
+ * handles expansions related to SObjectType. There could be exclusions that resolve to value types
+ * that hasn't been handled yet. TODO: Expand this to cover more chained name prefixes to improve
+ * readability in Factory classes.
+ */
+public final class SObjectTypeChainUtil {
```
Comment (rmohan20) [inline sfge/src/main/java/com/salesforce/graph/symbols/apex/schema/SObjectTypeChainUtil.java:15]: New utility to handle some chained names redundant logic. I'm restricting it to only `SObjectType` related chained names that are mostly related to `DescribeSObjectResult` factory. Wary of the side effects if I expand it too much. I've also added more test cases to various schema-related apex value tests to make sure this hasn't broken any of the existing formats.

File: sfge/src/main/java/com/salesforce/graph/symbols/apex/schema/SObjectFieldFactory.java [hunk 7]
```diff
@@ -12,36 +14,51 @@
 @SuppressWarnings(
         "PMD.RedundantFieldInitializer") // SCHEMA_INDEX is intentionally zero to provide context
 public final class SObjectFieldFactory {
```
Comment (rmohan20) [inline sfge/src/main/java/com/salesforce/graph/symbols/apex/schema/SObjectFieldFactory.java:16]: Not bringing the chained names handling to fields at the moment. This brings in a lot more cases that need a lot of testing effort. Will revisit at the next opportunity.

File: sfge/src/main/java/com/salesforce/graph/symbols/apex/schema/DescribeSObjectResultFactory.java [hunk 8]
```diff
@@ -3,136 +3,170 @@
...
+                        // Schema.SObjectType.Account.isDeletable()
+                        // SObjectType.Account.isDeletable()
+                        return getDmlAccessValue(vertex, sObjectName.get());
                     }
                 }
+
                 return Optional.empty();
             };
 
+    /**
```
Comment (rmohan20) [inline sfge/src/main/java/com/salesforce/graph/symbols/apex/schema/DescribeSObjectResultFactory.java:109]: All of this logic already existed. I've only extracted them into individual methods for reusability.

File: sfge/src/main/java/com/salesforce/graph/symbols/apex/schema/DescribeSObjectResultFactory.java [hunk 9]
```diff
@@ -3,136 +3,170 @@
...
+                        // TODO: how do we handle other fields with individual types?
+                        //  Stretch goal when we get to parsing field definitions of SObjects
+                        return getApexStringValue(sObjectName.get());
+                    } else if (ApexStandardLibraryUtil.VariableNames.S_OBJECT_TYPE.equalsIgnoreCase(
+                            vertexName)) {
+                        // Schema.SObjectType.Account.SObjectType is a SObjectType value.
+                        //  Reject this as a {@link DescribeSObjectResult}
+                        // TODO: streamline all of this
+                        return Optional.empty();
                     }
```
Comment (rmohan20) [inline sfge/src/main/java/com/salesforce/graph/symbols/apex/schema/DescribeSObjectResultFactory.java:59]: These are the kind of weird outliers that I'm wary of - glad we had a test to detect this. Added a skip case.

**Review by jfeingold35 — APPROVED** [review]

**Pull request merged**